cirq-core/cirq/sim/clifford/act_on_stabilizer_args.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+from typing import Any, Sequence, TYPE_CHECKING, Union
+
+import numpy as np
+
+from cirq import ops, protocols, linalg
+from cirq.ops import common_gates, matrix_gates, global_phase_op
+from cirq.ops.clifford_gate import SingleQubitCliffordGate
+from cirq.protocols import has_unitary, num_qubits, unitary
+from cirq.sim.act_on_args import ActOnArgs
+from cirq.type_workarounds import NotImplementedType
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class ActOnStabilizerArgs(ActOnArgs, metaclass=abc.ABCMeta):
+    """Abstract wrapper around a stabilizer state for the act_on protocol."""
+
+    def _act_on_fallback_(
+        self,
+        action: Union['cirq.Operation', 'cirq.Gate'],
+        qubits: Sequence['cirq.Qid'],
+        allow_decompose: bool = True,
+    ) -> Union[bool, NotImplementedType]:
+        strats = [
+            self._strat_apply_gate,
+            self._strat_apply_mixture,
+            self._strat_decompose,
```

Comment:
This allows `_strat_decompose` even if `allow_decompose` is false.

Reply:
Yeah, putting this into the `allow_decompose` branch caused some of @ybc1991 's tests to fail. In particular, when testing `_strat_act_from_single_qubit_decompose`, those tests mostly use `allow_decompose=False`, (e.g. https://github.com/quantumlib/Cirq/blob/efa082051ebb6b490567a7753884c33c1a0f9666/cirq-core/cirq/transformers/analytical_decompositions/clifford_decomposition_test.py#L140). 

The reason it fails is that the unitary decomp has SWAP as a possible result of the decomposition. However, SWAP is not a gate that I'm handling natively here, instead allowing it to fall back to decomposing into "CX01 CX10 CX01". However since that particular test sets `allow_decompose=False` the `_strat_decompose` doesn't run if it's not in the main strats list.

Nonetheless you're right it shouldn't be in the main strats list. The two options would be either for those tests to run with `allow_decompose=True`, or to handle SWAP gates explicitly here rather than falling back to decompose. I lean toward the latter. (I could even provide the default implementation that calls CX 3x). Sound good?

I'll also have to add an explicit decomposition test after that, as SWAP was the thing that implicitly tested the decomposition logic.

Reply:
I fixed this by handling SWAP explicitly.